pkg/controller/jobs/statefulset/statefulset_reconciler.go

```diff
@@ -0,0 +1,211 @@
+	var updated bool
+
+	if shouldUngate(sts, pod) && utilpod.Ungate(pod, podcontroller.SchedulingGateName) {
+		updated = true
+	}
+
+	if shouldFinalize(sts, pod) && controllerutil.RemoveFinalizer(pod, podcontroller.PodFinalizer) {
+		updated = true
+	}
+
+	return updated
+}
+
+func shouldUngate(sts *appsv1.StatefulSet, pod *corev1.Pod) bool {
+	return sts == nil || sts.Status.CurrentRevision != sts.Status.UpdateRevision &&
+		sts.Status.CurrentRevision == pod.Labels[appsv1.ControllerRevisionHashLabelKey]
+}
+
+func shouldFinalize(sts *appsv1.StatefulSet, pod *corev1.Pod) bool {
+	return shouldUngate(sts, pod) || pod.Status.Phase == corev1.PodSucceeded || pod.Status.Phase == corev1.PodFailed
+}
+
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	ctrl.Log.V(3).Info("Setting up StatefulSet reconciler")
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&appsv1.StatefulSet{}).
+		WithEventFilter(r).
+		Watches(&corev1.Pod{}, &podHandler{}).
+		Complete(r)
+}
+
+func NewReconciler(client client.Client, _ record.EventRecorder, _ ...jobframework.Option) jobframework.JobReconcilerInterface {
+	return &Reconciler{client: client}
+}
+
+var _ predicate.Predicate = (*Reconciler)(nil)
+
+func (r *Reconciler) Generic(event.GenericEvent) bool {
+	return false
+}
+
+func (r *Reconciler) Create(e event.CreateEvent) bool {
+	return r.handle(e.Object)
+}
+
+func (r *Reconciler) Update(e event.UpdateEvent) bool {
+	return r.handle(e.ObjectNew)
+}
+
+func (r *Reconciler) Delete(e event.DeleteEvent) bool {
+	return r.handle(e.Object)
+}
+
+func (r *Reconciler) handle(obj client.Object) bool {
+	sts, isSts := obj.(*appsv1.StatefulSet)
+	if !isSts {
+		return true
+	}
+	// Handle only statefulset managed by kueue.
+	return isManagedByKueue(sts)
+}
+
+var _ handler.EventHandler = (*podHandler)(nil)
+
+type podHandler struct{}
+
+func (h *podHandler) Generic(context.Context, event.GenericEvent, workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+}
+
+func (h *podHandler) Create(_ context.Context, e event.CreateEvent, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	// To watch for pod ADDED events.
+	h.handle(e.Object, q)
+}
+
+func (h *podHandler) Update(_ context.Context, e event.UpdateEvent, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+}
+
+func (h *podHandler) Delete(context.Context, event.DeleteEvent, workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+}
+
+func (h *podHandler) handle(obj client.Object, q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod {
+		return
+	}
+	if controllerRef := metav1.GetControllerOf(pod); controllerRef != nil {
+		if controllerRef.Kind != gvk.Kind || controllerRef.APIVersion != gvk.GroupVersion().String() {
+			// The pod is controlled by an owner that is not an apps/v1 StatefulSet.
+			return
+		}
+		if _, ok := pod.Labels[podcontroller.GroupNameLabel]; !ok {
+			return
+		}
```

Comment:
Add a comment:
```
// The check for the "group-name" label is introduced specifically for the 0.9
// branch to avoid reconciling all StatefulSets, as the 0.9 branch does not yet
// have the `pod-suspending-parent` annotation.
```

Reply:
Thanks! Added.